File: Online Judges/LeetCode/June LeetCoding Challenge/Week-4/Perfect Squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [1e9]*(n+1)
        dp[0] = 0
        for i in range(1, n+1):
            for j in range(1, i+1):
                if(j*j > i):
                    break
                dp[i] = min(dp[i], dp[i-j*j] + 1)
        return dp[n]

File: Online Judges/LeetCode/June LeetCoding Challenge/Week-4/Perfect Squares.rs
use std::cmp;
impl Solution {
    pub fn num_squares(n: i32) -> i32 {
        let n = n as usize;
        let mut dp = vec![1009; n+1];
        dp[0] = 0;
        for i in 1..n+1 {
            for j in 1..i+1 {
                if j*j<=i {
                    dp[i] = cmp::min(dp[i], dp[i-j*j] +1);
                }
            }
        }
        dp[n]
    }
}